`non_zero_suggestions` suggestion on signed `/` and `%` auto-fixes to code that will not compile

### Summary

`non_zero_suggestions` suggests using NonZeroI64 in cases where it's not actually possible to do so, and auto-fixes code such that it will not successfully compile.

### Lint Name

non_zero_suggestions

### Reproducer

I tried this code:

```rust
use std::num::NonZeroI32;

fn main() {
    let s: i64 = 10;
    let t = NonZeroI32::new(3).unwrap();

    let quotient = s / i64::from(t.get());
    let remainder = s % i64::from(t.get());

    println!("{quotient} {remainder}");
}
```

I saw this happen:

```
warning: consider using `NonZeroI64::from()` for more efficient and type-safe conversion
 --> src/main.rs:7:24
  |
7 |     let quotient = s / i64::from(t.get());
  |                        ^^^^^^^^^^^^^^^^^^ help: replace with: `NonZeroI64::from(t)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.96.0/index.html#non_zero_suggestions
  = note: requested on the command line with `-W clippy::non-zero-suggestions`

warning: consider using `NonZeroI64::from()` for more efficient and type-safe conversion
 --> src/main.rs:8:25
  |
8 |     let remainder = s % i64::from(t.get());
  |                         ^^^^^^^^^^^^^^^^^^ help: replace with: `NonZeroI64::from(t)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.96.0/index.html#non_zero_suggestions

warning: `repro-clippy-issue` (bin "repro-clippy-issue") generated 2 warnings (run `cargo clippy --fix --bin "repro-clippy-issue" -p repro-clippy-issue -- ` to apply 2 suggestions)
```

I expected to see this happen:

This particular code would not result in suggestions with this rule.

`/` and `%` aren't implemented on non-zero signed integers, and so this lint suggestion results in code that will fail to compile. The autofix creates invalid code here.

I get this error after running the autofix:
```
= cause: error[E0277]: cannot divide `i64` by `std::num::NonZero<i64>`
              --> src/main.rs:9:22
               |
             9 |     let quotient = s / NonZeroI64::from(t);
               |                      ^ no implementation for `i64 / std::num::NonZero<i64>`
               |
               = help: the trait `std::ops::Div<std::num::NonZero<i64>>` is not implemented for `i64`
           help: the following other types implement trait `std::ops::Div<Rhs>`
              --> /Users/connorshea/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/arith.rs:488:9
               |
           488 |           impl const Div for $t {
               |           ^^^^^^^^^^^^^^^^^^^^^ `i64` implements `std::ops::Div`
           ...
           502 | / div_impl_integer! {
           503 | |     (usize u8 u16 u32 u64 u128) => "This operation will panic if `other == 0`.",
           504 | |     (isize i8 i16 i32 i64 i128) => "This operation will panic if `other == 0` or the division results in overflow."
           505 | | }
               | |_- in this macro invocation
               |
              ::: /Users/connorshea/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/internal_macros.rs:22:9
               |
            22 |           impl const $imp<$u> for &$t {
               |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `&i64` implements `std::ops::Div<i64>`
           ...
            33 |           impl const $imp<&$u> for $t {
               |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `i64` implements `std::ops::Div<&i64>`
           ...
            44 |           impl const $imp<&$u> for &$t {
               |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `&i64` implements `std::ops::Div`
               = note: this error originates in the macro `div_impl_integer` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Version

```text
rustc 1.96.1 (31fca3adb 2026-06-26)
binary: rustc
commit-hash: 31fca3adb283cc9dfd56b49cdee9a96eb9c96ffd
commit-date: 2026-06-26
host: aarch64-apple-darwin
release: 1.96.1
LLVM version: 22.1.2
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

--- a/clippy_lints/src/non_zero_suggestions.rs
+++ b/clippy_lints/src/non_zero_suggestions.rs
@@ -52,7 +52,13 @@ impl<'tcx> LateLintPass<'tcx> for NonZeroSuggestions {
         if let ExprKind::Binary(op, _, rhs) = expr.kind
             && matches!(op.node, BinOpKind::Div | BinOpKind::Rem)
         {
-            check_non_zero_conversion(cx, rhs, Applicability::MachineApplicable);
+            // `Div<NonZero<T>>` and `Rem<NonZero<T>>` are only implemented for unsigned
+            // integer types; signed integer types do not have these impls, so suggesting
+            // `NonZeroI*::from(x)` as the divisor produces code that won't compile.
+            let rhs_ty = cx.typeck_results().expr_ty(rhs);
+            if matches!(rhs_ty.kind(), ty::Uint(_)) {
+                check_non_zero_conversion(cx, rhs, Applicability::MachineApplicable);
+            }
         } else {
             // Check if the parent expression is a binary operation
             let parent_is_binary = cx.tcx.hir_parent_iter(expr.hir_id).any(|(_, node)| {

--- a/tests/ui/non_zero_suggestions.fixed
+++ b/tests/ui/non_zero_suggestions.fixed
@@ -1,5 +1,5 @@
 #![warn(clippy::non_zero_suggestions)]
-use std::num::{NonZeroI8, NonZeroI16, NonZeroU8, NonZeroU16, NonZeroU32, NonZeroU64, NonZeroUsize};
+use std::num::{NonZeroI8, NonZeroI16, NonZeroI32, NonZeroU8, NonZeroU16, NonZeroU32, NonZeroU64, NonZeroUsize};
 
 fn main() {
     /// Positive test cases (lint should trigger)
@@ -45,6 +45,12 @@ fn main() {
     let k: u64 = 300;
     let l = NonZeroU32::new(15).unwrap();
     let r9 = k / NonZeroU64::from(l);
+
+    // Signed integers in div/rem: must NOT trigger because i64 / NonZeroI64 is not in std
+    let s: i64 = 10;
+    let t = NonZeroI32::new(3).unwrap();
+    let r10 = s / i64::from(t.get());
+    let r11 = s % i64::from(t.get());
 }
 
 // Additional function to test the lint in a different context

--- a/tests/ui/non_zero_suggestions.rs
+++ b/tests/ui/non_zero_suggestions.rs
@@ -1,5 +1,5 @@
 #![warn(clippy::non_zero_suggestions)]
-use std::num::{NonZeroI8, NonZeroI16, NonZeroU8, NonZeroU16, NonZeroU32, NonZeroU64, NonZeroUsize};
+use std::num::{NonZeroI8, NonZeroI16, NonZeroI32, NonZeroU8, NonZeroU16, NonZeroU32, NonZeroU64, NonZeroUsize};
 
 fn main() {
     /// Positive test cases (lint should trigger)
@@ -45,6 +45,12 @@ fn main() {
     let k: u64 = 300;
     let l = NonZeroU32::new(15).unwrap();
     let r9 = k / NonZeroU64::from(l);
+
+    // Signed integers in div/rem: must NOT trigger because i64 / NonZeroI64 is not in std
+    let s: i64 = 10;
+    let t = NonZeroI32::new(3).unwrap();
+    let r10 = s / i64::from(t.get());
+    let r11 = s % i64::from(t.get());
 }
 
 // Additional function to test the lint in a different context

--- a/tests/ui/non_zero_suggestions.stderr
+++ b/tests/ui/non_zero_suggestions.stderr
@@ -26,13 +26,13 @@ LL |     let x = u64::from(NonZeroU32::new(5).unwrap().get());
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `NonZeroU64::from(NonZeroU32::new(5).unwrap())`
 
 error: consider using `NonZeroU64::from()` for more efficient and type-safe conversion
-  --> tests/ui/non_zero_suggestions.rs:52:9
+  --> tests/ui/non_zero_suggestions.rs:58:9
    |
 LL |     x / u64::from(y.get())
    |         ^^^^^^^^^^^^^^^^^^ help: replace with: `NonZeroU64::from(y)`
 
 error: consider using `NonZeroU64::from()` for more efficient and type-safe conversion
-  --> tests/ui/non_zero_suggestions.rs:62:22
+  --> tests/ui/non_zero_suggestions.rs:68:22
    |
 LL |         self.value / u64::from(divisor.get())
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `NonZeroU64::from(divisor)`

